component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.tlswithidsecret/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/tlswithidsecret/MutualTLSWithIdSecretAuthenticator.java

```diff
@@ -0,0 +1,223 @@
+                        if (log.isDebugEnabled()) {
+                            log.debug("Error while retrieving configured certificate.", e);
+                            log.debug("Client certificate is not configured for the app with client id: " +
+                                    oAuthClientAuthnContext.getClientId() + ". Therefore not validating cert");
+                        }
+                        return true;
+                    }
+                }
+            }
+
+            if (log.isDebugEnabled()) {
+                log.debug("Authenticating client : " + oAuthClientAuthnContext.getClientId() + " with public " +
+                        "certificate.");
+            }
+
+            X509Certificate requestCert;
+            Object certObject = request.getAttribute(JAVAX_SERVLET_REQUEST_CERTIFICATE);
+            if (certObject instanceof X509Certificate[]) {
+                X509Certificate[] cert = (X509Certificate[]) certObject;
+                requestCert = cert[0];
+            } else if (certObject instanceof X509Certificate) {
+                requestCert = (X509Certificate) certObject;
+            } else {
+                log.error("Could not find client certificate in required format in the request for client: " +
+                        oAuthClientAuthnContext.getClientId());
+                return false;
+            }
+
+            return authenticate(registeredCert, requestCert);
+
+        } catch (IdentityOAuth2Exception e) {
+            throw new OAuthClientAuthnException(OAuth2ErrorCodes.SERVER_ERROR, "Error occurred while retrieving " +
+                    "public certificate of client ID: " + oAuthClientAuthnContext.getClientId(), e);
+        } catch (InvalidOAuthClientException e) {
+            throw new OAuthClientAuthnException(OAuth2ErrorCodes.INVALID_CLIENT, "Error occurred while retrieving " +
+                    "tenant domain for the client ID: " + oAuthClientAuthnContext.getClientId(), e);
+        } catch (Exception e) {
+            throw new OAuthClientAuthnException(OAuth2ErrorCodes.SERVER_ERROR,
+                    "Unexpected error while authenticating client: " + oAuthClientAuthnContext.getClientId(), e);
+        }
+
+    }
+
+    private String getMandateMutualSSLProperty() {
+
+        return IdentityUtil.readEventListenerProperty
+                (AbstractIdentityHandler.class.getName(), this.getClass().getName()).getProperties().
+                getProperty(MANDATE_MUTUAL_SSL);
+    }
+
+    public boolean canAuthenticate(HttpServletRequest request, Map<String, List> bodyParams,
+                                   OAuthClientAuthnContext oAuthClientAuthnContext) {
+
+        // This authenticator will be skipped if BasicAuthClientAuthenticator was engaged
+        return !oAuthClientAuthnContext.getExecutedAuthenticators().contains(super.getName()) &&
+                super.canAuthenticate(request, bodyParams, oAuthClientAuthnContext);
```

Comment:
Did we check the behavior when other client authenticators also evaluate canAuthenticate to true? e.g. MutualTLS authenticator. IMO we should have a way to uniquely engage this authenticator in such cases.
https://github.com/wso2-extensions/identity-oauth-addons/blob/master/component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.mutualtls/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/mutualtls/MutualTLSClientAuthenticator.java#L176

Reply:
Regarding other client authenticators, the `MutualTLSWithIdSecretAuthenticator` would evaluate `false` for `canhandle()` only if `BasicAuthClientAuthenticator` is there in the executed authenticators list. So, it would work along with the MutualTLS authenticator.